napari/layers/_shapes_layer/model.py

```diff
@@ -0,0 +1,1521 @@
+                    if shift and shape is not None:
+                        if shape in self.selected_shapes:
+                            self.selected_shapes.remove(shape)
+                            shapes = self.selected_shapes
+                            self._selected_box = self.bounding_box(shapes)
+                        else:
+                            self.selected_shapes.append(shape)
+                            shapes = self.selected_shapes
+                            self._selected_box = self.bounding_box(shapes)
+                    elif shape is not None:
+                        if shape not in self.selected_shapes:
+                            self.selected_shapes = [shape]
+                    else:
+                        self.selected_shapes = []
+                self._set_highlight()
+                self.status = self.get_message(coord, shape, vertex)
+        elif self.mode == 'direct':
+            if not self._is_moving and not self._is_selecting:
+                shape, vertex = self._shape_at(coord)
+                self._moving_shape = shape
+                self._moving_vertex = vertex
+                if vertex is None:
+                    if shift and shape is not None:
+                        if shape in self.selected_shapes:
+                            self.selected_shapes.remove(shape)
+                            shapes = self.selected_shapes
+                            self._selected_box = self.bounding_box(shapes)
+                        else:
+                            self.selected_shapes.append(shape)
+                            shapes = self.selected_shapes
+                            self._selected_box = self.bounding_box(shapes)
+                    elif shape is not None:
+                        if shape not in self.selected_shapes:
+                            self.selected_shapes = [shape]
+                    else:
+                        self.selected_shapes = []
+                self._set_highlight()
+                self.status = self.get_message(coord, shape, vertex)
+        elif (self.mode == 'add_rectangle' or self.mode == 'add_ellipse' or
+              self.mode == 'add_line'):
+            # Start drawing a rectangle / ellipse / line
+            transform = self.viewer._canvas.scene.node_transform(self._node)
+            rescale = transform.map([1, 1])[:2] - transform.map([0, 0])[:2]
+            size = self._vertex_size*rescale.mean()/4
+            new_z_index = max(self.data._z_index, default=-1) + 1
+            if self.mode == 'add_rectangle':
+                data = np.array([coord, coord+size])
+                self.data.add(data, shape_type='rectangle',
+                              edge_width=self.edge_width,
+                              edge_color=self.edge_color,
+                              face_color=self.face_color,
+                              opacity=self.opacity,
+                              z_index=new_z_index)
```

Comment:
You could define `draw_attrs = dict(edge_width=self.edge_width, ...)` then `self.data.add(data, shape_type='rectangle', **draw_attrs)`

Reply:
I changed some control flow logic so this was no longer duplicated, and so decided against a dict